gobblin-yarn/src/main/java/org/apache/gobblin/yarn/GobblinApplicationMaster.java

```diff
@@ -104,7 +104,9 @@ public GobblinApplicationMaster(String applicationName, String applicationId, Co
 
     if (UserGroupInformation.isSecurityEnabled()) {
       LOGGER.info("Adding YarnContainerSecurityManager since security is enabled");
-      this.applicationLauncher.addService(buildYarnContainerSecurityManager(this.config, this.fs));
+      YarnContainerSecurityManager yarnContainerSecurityManager = buildYarnContainerSecurityManager(this.config, this.fs);
```

Comment:
We need to be a little careful on when we add the tokens on AM restart. In particular, we need to ensure that the new tokens are added into the UGI before any HDFS/Hive/RM access happens. Your fix may be ok, but in the future, this detail may easily be forgotten and may introduce unintentional bugs. 
Do you think it would be safer to add the credentials inside the main() of the GobblinApplicationMaster? Also, instead of sending a DelegationTokenUpdateEvent, a cleaner approach maybe to make the readCredentials and addCredentials methods public static and call them directly in the main?

Reply:
One more comment: For safety, it may make sense to do the same action on the worker nodes as well i.e. on start up read the latest tokens from the token file.